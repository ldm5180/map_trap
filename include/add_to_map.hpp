#ifndef ADD_TO_MAP_HPP
#define ADD_TO_MAP_HPP

#include <algorithm>
#include <benchmark/benchmark.h>
#include <cstdlib>
#include <string>

static const auto add_to_map = [](const auto num_elems, auto &map) {
  // Add elements to map, this sorts automatically.
  for (auto i = 0; i < num_elems; ++i) {
    const auto random_key = std::rand();
    map[random_key] = std::to_string(map.size());
  }
  benchmark::ClobberMemory();
};

static const auto add_to_map_and_key_list = [](const auto num_elems, auto &map,
                                               auto &keys) {
  // Add elements to map, this sorts automatically.
  for (auto i = 0; i < num_elems; ++i) {
    const auto random_key = std::rand();
    map[random_key] = std::to_string(map.size());
    keys.push_back(random_key);
    std::sort(keys.begin(), keys.end());
  }
  benchmark::ClobberMemory();
};

#endif // ADD_TO_MAP_HPP
