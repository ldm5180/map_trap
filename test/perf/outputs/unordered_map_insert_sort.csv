Run on (4 X 2594.09 MHz CPU s)
2017-10-28 16:40:15
name,iterations,real_time,cpu_time,time_unit,bytes_per_second,items_per_second,label,error_occurred,error_message
"BM_unordered_map_insert_sort/8",1000,768672,768469,ns,,,,,
"BM_unordered_map_insert_sort/16",1000,3.43057e+06,3.42523e+06,ns,,,,,
"BM_unordered_map_insert_sort/32",612,8.76875e+06,8.76754e+06,ns,,,,,
"BM_unordered_map_insert_sort/64",146,7.64083e+06,7.64077e+06,ns,,,,,
"BM_unordered_map_insert_sort/128",100,2.13776e+07,2.13765e+07,ns,,,,,
"BM_unordered_map_insert_sort/256",99,9.19262e+07,9.19101e+07,ns,,,,,
"BM_unordered_map_insert_sort/512",23,7.71712e+07,7.71451e+07,ns,,,,,
"BM_unordered_map_insert_sort/1024",10,1.27079e+08,1.27074e+08,ns,,,,,
"BM_unordered_map_insert_sort/2048",10,5.95856e+08,5.95786e+08,ns,,,,,
