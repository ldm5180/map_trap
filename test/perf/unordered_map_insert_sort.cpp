#include "add_to_map.hpp"

#include <benchmark/benchmark.h>
#include <cstdlib>
#include <string>
#include <unordered_map>

using map_t = std::unordered_map<int, std::string>;

static void BM_unordered_map_insert_sort(benchmark::State &state) {
  std::srand(42);
  map_t map{};
  map.reserve(state.range(0));
  std::vector<map_t::key_type> keys{};
  keys.reserve(state.range(0));

  for (auto _ : state) {
    add_to_map_and_key_list(state.range(0), map, keys);
  }
}
BENCHMARK(BM_unordered_map_insert_sort)->RangeMultiplier(2)->Range(8, 8 << 8);
BENCHMARK_MAIN();
