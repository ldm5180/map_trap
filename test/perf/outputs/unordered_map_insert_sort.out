Run on (4 X 2594.09 MHz CPU s)
2017-10-28 16:28:38
name,iterations,real_time,cpu_time,time_unit,bytes_per_second,items_per_second,label,error_occurred,error_message
"BM_unordered_map_insert_sort/8",1000,774967,772184,ns,,,,,
"BM_unordered_map_insert_sort/16",1000,3.37564e+06,3.3763e+06,ns,,,,,
"BM_unordered_map_insert_sort/32",613,8.6373e+06,8.63991e+06,ns,,,,,
"BM_unordered_map_insert_sort/64",144,7.07123e+06,7.07458e+06,ns,,,,,
"BM_unordered_map_insert_sort/128",100,2.10116e+07,2.10173e+07,ns,,,,,
"BM_unordered_map_insert_sort/256",105,9.78398e+07,9.7867e+07,ns,,,,,
"BM_unordered_map_insert_sort/512",23,7.65591e+07,7.65751e+07,ns,,,,,
"BM_unordered_map_insert_sort/1024",10,1.28467e+08,1.28498e+08,ns,,,,,
"BM_unordered_map_insert_sort/2048",10,5.75048e+08,5.75199e+08,ns,,,,,
